### PR TITLE
feat(composer): dynamic installation

### LINKED ISSUE #8497: Support configurable PHP version when binarySource=docker
**What would you like Renovate to be able to do?**

Support configurable PHP version when running `composer`.

**Did you already have any implementation ideas?**

Use `constraints.php` in config to pick a `renovate/php` version, which are already pre-built. Then dynamically install `composer` at runtime into user space as a pre-command.

### PR DESCRIPTION
**Pull request by viceice**
## Changes:

Install composer dynamically on php docker image, so we can respect php contraints.

https://github.com/viceice-tests/renovate-composer/pull/7

## Context:

closes #8497

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by viceice** [comment]
Debug logs:
```
DEBUG: Using composer constraint (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
       "range": "~2.0.0",
       "version": "2.0.14"
DEBUG: Using php constraint from config (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
DEBUG: composer command (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
       "cmd": "composer",
       "args": "update doctrine/migrations --with-dependencies --ignore-platform-reqs --no-ansi --no-interaction --no-scripts --no-autoloader --no-plugins"
DEBUG: Using docker to execute (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
DEBUG: Found version constraint - checking for a compatible image to use (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
       "depName": "docker.io/renovate/php",
       "scheme": "composer",
       "constraint": "7.3"
DEBUG: getLabels(https://index.docker.io, renovate/php, latest) (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
DEBUG: getManifestResponse(https://index.docker.io, renovate/php, latest) (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
DEBUG: found labels in manifest (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
       "labels": {
         "maintainer": "Rhys Arkins <rhys@arkins.net>",
         "org.opencontainers.image.source": "https://github.com/renovatebot/docker-php",
         "org.opencontainers.image.version": "7.4.22"
       }
DEBUG: Found compatible image version (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
       "depName": "docker.io/renovate/php",
       "scheme": "composer",
       "constraint": "7.3",
       "version": "7.3"
DEBUG: Resolved tag constraint (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
       "image": "docker.io/renovate/php",
       "tagConstraint": "7.3",
       "tagVersioning": "composer",
       "tag": "7.3"
DEBUG: Fetching Docker image: docker.io/renovate/php:7.3 (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
DEBUG: Finished fetching Docker image (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
DEBUG: Executing command (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
       "command": "docker run --rm --name=renovate_php --label=renovate_child -v \"/tmp/renovate/repos/github/viceice-tests/renovate-composer\":\"/tmp/renovate/repos/github/viceice-tests/renovate-composer\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e COMPOSER_CACHE_DIR -e COMPOSER_AUTH -e TEST_KEY -e a -w \"/tmp/renovate/repos/github/viceice-tests/renovate-composer\" docker.io/renovate/php:7.3 bash -l -c \"install-tool composer 2.0.14 && composer update doctrine/migrations --with-dependencies --ignore-platform-reqs --no-ansi --no-interaction --no-scripts --no-autoloader --no-plugins\""
DEBUG: exec completed (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
       "cmd": "docker run --rm --name=renovate_php --label=renovate_child -v \"/tmp/renovate/repos/github/viceice-tests/renovate-composer\":\"/tmp/renovate/repos/github/viceice-tests/renovate-composer\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e COMPOSER_CACHE_DIR -e COMPOSER_AUTH -e TEST_KEY -e a -w \"/tmp/renovate/repos/github/viceice-tests/renovate-composer\" docker.io/renovate/php:7.3 bash -l -c \"install-tool composer 2.0.14 && composer update doctrine/migrations --with-dependencies --ignore-platform-reqs --no-ansi --no-interaction --no-scripts --no-autoloader --no-plugins\"",
       "durationMs": 3978,
       "stdout": "Installing tool composer v2.0.14\nComposer version 2.0.14 2021-05-21 17:03:37\n",
       "stderr": "Loading composer repositories with package information\nUpdating dependencies\nLock file operations: 2 installs, 18 updates, 2 removals\n  - Removing laminas/laminas-eventmanager (3.2.1)\n  - Removing laminas/laminas-zendframework-bridge (1.0.4)\n  - Upgrading doctrine/cache (1.10.2 => 2.1.1)\n  - Upgrading doctrine/dbal (2.10.2 => 2.13.2)\n  - Locking doctrine/deprecations (v0.5.3)\n  - Upgrading laminas/laminas-code (3.4.1 => 4.4.2)\n  - Upgrading ocramius/package-versions (1.9.0 => 1.11.0)\n  - Upgrading ocramius/proxy-manager (2.8.1 => 2.13.0)\n  - Upgrading psr/container (1.0.0 => 1.1.1)\n  - Upgrading psr/log (1.1.3 => 1.1.4)\n  - Upgrading symfony/console (v5.1.3 => v5.3.6)\n  - Locking symfony/deprecation-contracts (v2.4.0)\n  - Upgrading symfony/polyfill-ctype (v1.18.1 => v1.23.0)\n  - Upgrading symfony/polyfill-intl-grapheme (v1.18.1 => v1.23.1)\n  - Upgrading symfony/polyfill-intl-normalizer (v1.18.1 => v1.23.0)\n  - Upgrading symfony/polyfill-mbstring (v1.18.1 => v1.23.1)\n  - Upgrading symfony/polyfill-php73 (v1.18.1 => v1.23.0)\n  - Upgrading symfony/polyfill-php80 (v1.18.1 => v1.23.1)\n  - Upgrading symfony/service-contracts (v2.1.3 => v2.4.0)\n  - Upgrading symfony/stopwatch (v5.1.3 => v5.3.4)\n  - Upgrading symfony/string (v5.1.3 => v5.3.3)\n  - Upgrading webimpress/safe-writer (2.0.1 => 2.2.0)\nWriting lock file\nInstalling dependencies from lock file (including require-dev)\nPackage operations: 22 installs, 0 updates, 0 removals\n  - Installing ocramius/package-versions (1.11.0): Extracting archive\n  - Installing doctrine/cache (2.1.1): Extracting archive\n  - Installing doctrine/deprecations (v0.5.3): Extracting archive\n  - Installing psr/container (1.1.1): Extracting archive\n  - Installing symfony/service-contracts (v2.4.0): Extracting archive\n  - Installing symfony/stopwatch (v5.3.4): Extracting archive\n  - Installing symfony/polyfill-php80 (v1.23.1): Extracting archive\n  - Installing symfony/polyfill-mbstring (v1.23.1): Extracting archive\n  - Installing symfony/polyfill-intl-normalizer (v1.23.0): Extracting archive\n  - Installing symfony/polyfill-intl-grapheme (v1.23.1): Extracting archive\n  - Installing symfony/polyfill-ctype (v1.23.0): Extracting archive\n  - Installing symfony/string (v5.3.3): Extracting archive\n  - Installing symfony/polyfill-php73 (v1.23.0): Extracting archive\n  - Installing symfony/deprecation-contracts (v2.4.0): Extracting archive\n  - Installing symfony/console (v5.3.6): Extracting archive\n  - Installing psr/log (1.1.4): Extracting archive\n  - Installing webimpress/safe-writer (2.2.0): Extracting archive\n  - Installing laminas/laminas-code (4.4.2): Extracting archive\n  - Installing ocramius/proxy-manager (2.13.0): Extracting archive\n  - Installing doctrine/event-manager (1.1.1): Extracting archive\n  - Installing doctrine/dbal (2.13.2): Extracting archive\n  - Installing doctrine/migrations (3.0.1): Extracting archive\n19 packages you are using are looking for funding.\nUse the `composer fund` command to find out more!\n"
DEBUG: Returning updated composer.lock (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
DEBUG: Updated 1 package files (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
DEBUG: Updated 1 lock files (repository=viceice-tests/renovate-composer, branch=renovate-test/pin-dependencies)
       "updatedArtifacts": ["composer.lock"]
```

File: lib/manager/composer/utils.ts [hunk 1]
```diff
@@ -1,17 +1,57 @@
...
+  let versions = release.releases.map((r) => r.version);
+
+  if (composer) {
+    versions = versions.filter(
+      (v) => api.isValid(v) && api.matches(v, composer)
+    );
+  }
+
+  if (!versions.length) {
+    throw new Error('No compatible composer releases found.');
```
Comment (rarkins) [inline lib/manager/composer/utils.ts:39]: Any way to fall back to latest? Or best we abort?

File: lib/manager/composer/__snapshots__/artifacts.spec.ts.snap [hunk 2]
```diff
@@ -121,19 +121,19 @@ Array [
 exports[`.updateArtifacts() supports docker mode 1`] = `
 Array [
   Object {
-    "cmd": "docker pull renovate/composer:1.10.17",
+    "cmd": "docker pull renovate/php:latest",
```
Comment (viceice) [inline lib/manager/composer/__snapshots__/artifacts.spec.ts.snap:124]: 🤔  should use php 7.3, need to fix the test

File: lib/manager/composer/utils.ts (hunk 1 above)
Comment (viceice) [reply lib/manager/composer/utils.ts:39]: Abort, as this must be a temporary error.

File: lib/manager/composer/__snapshots__/artifacts.spec.ts.snap (hunk 2 above)
Comment (viceice) [reply lib/manager/composer/__snapshots__/artifacts.spec.ts.snap]: fixed

File: lib/manager/composer/__snapshots__/artifacts.spec.ts.snap [hunk 3]
```diff
@@ -121,19 +121,19 @@ Array [
 exports[`.updateArtifacts() supports docker mode 1`] = `
```
Comment (HonkingGoose) [inline lib/manager/composer/__snapshots__/artifacts.spec.ts.snap:121]: Would a PR title like `feat(composer): dynamic installation` or `feat(composer): dynamic install` be better.

`instal` is not spelt correctly in any case. 😉

**Review by rarkins — APPROVED** [review]

**Pull request merged**